rust/gtk-client/src/main.rs

```diff
@@ -0,0 +1,384 @@
+            .application(app)
+            .default_width(640)
+            .default_height(480)
+            .title("Firezone GTK+ 3")
+            .build();
+
+        // Don't forget to make all widgets visible.
+        win.show_all();
+    });
+
+    gtk::init()?;
+    let tray_icon = TrayIconBuilder::new()
+        .with_tooltip(TOOLTIP)
+        .with_icon(icon_to_native_icon(&Icon::default()))
+        .build()?;
+
+    let (main_tx, main_rx) = mpsc::channel(100);
+
+    let l = MainThreadLoop {
+        app: app.clone(),
+        last_icon_set: Default::default(),
+        main_rx,
+        tray_icon,
+    };
+    glib::spawn_future_local(l.run());
+
+    let (ctlr_tx, ctlr_rx) = mpsc::channel(100);
+
+    deep_link::register(current_exe)?;
+    tokio::spawn(accept_deep_links(deep_link_server, ctlr_tx.clone()));
+
+    {
+        let ctlr_tx = ctlr_tx.clone();
+        MenuEvent::set_event_handler(Some(move |event: MenuEvent| {
+            let Ok(event) = serde_json::from_str::<common::system_tray::Event>(&event.id.0) else {
+                tracing::error!("Couldn't parse system tray event");
+                return;
+            };
+            if let Err(error) = ctlr_tx.blocking_send(ControllerRequest::SystemTrayMenu(event)) {
+                tracing::error!(?error, "Couldn't send system tray event to Controller");
+            }
+        }));
+    }
+
+    let (_updates_tx, updates_rx) = mpsc::channel(1);
+
+    rt.spawn(run_controller(
+        main_tx,
+        ctlr_tx,
+        ctlr_rx,
+        log_filter_reloader,
+        telemetry,
+        updates_rx,
+    ));
+
+    if app.run() != 0.into() {
+        anyhow::bail!("GTK main loop returned non-zero exit code");
+    }
+    Ok(())
+}
+
+// Worker task to accept deep links from a named pipe forever
+///
+/// * `server` An initial named pipe server to consume before making new servers. This lets us also use the named pipe to enforce single-instance
+async fn accept_deep_links(mut server: deep_link::Server, ctlr_tx: CtlrTx) -> Result<()> {
+    loop {
+        match server.accept().await {
+            Ok(bytes) => {
+                let url = SecretString::from_str(
+                    std::str::from_utf8(bytes.expose_secret())
+                        .context("Incoming deep link was not valid UTF-8")?,
+                )
+                .context("Impossible: can't wrap String into SecretString")?;
+                // Ignore errors from this, it would only happen if the app is shutting down, otherwise we would wait
+                ctlr_tx
+                    .send(ControllerRequest::SchemeRequest(url))
+                    .await
+                    .ok();
+            }
+            Err(error) => tracing::error!(?error, "error while accepting deep link"),
+        }
+        // We re-create the named pipe server every time we get a link, because of an oddity in the Windows API.
+        server = deep_link::Server::new().await?;
+    }
+}
+
+struct MainThreadLoop {
```

Comment:
```suggestion
#[must_use]
struct MainThreadLoop {
```

Not that we'd ever forget but might as well.